hedera-node/hedera-schedule-service-impl/src/main/java/com/hedera/node/app/service/schedule/impl/WritableScheduleStoreImpl.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.schedule.impl;
+
+import com.hedera.hapi.node.base.ScheduleID;
+import com.hedera.hapi.node.base.Timestamp;
+import com.hedera.hapi.node.state.schedule.Schedule;
+import com.hedera.node.app.service.schedule.WritableScheduleStore;
+import com.hedera.node.app.service.schedule.impl.handlers.ScheduleUtility;
+import com.hedera.node.app.spi.state.WritableKVState;
+import com.hedera.node.app.spi.state.WritableStates;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Instant;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A writable store that wraps a writable key-value state and supports operations required to create or update
+ * schedule objects as a result of ScheduleCreate, ScheduleSign, or ScheduleDelete transactions.
+ */
+public class WritableScheduleStoreImpl extends ReadableScheduleStoreImpl implements WritableScheduleStore {
+    private static final String SCHEDULE_NULL_FOR_DELETE_MESSAGE =
+            "Request to delete null schedule ID cannot be fulfilled.";
+    private static final String SCHEDULE_MISSING_FOR_DELETE_MESSAGE =
+            "Schedule to be deleted, %1$s, not found in state.";
+    private final WritableKVState<ScheduleID, Schedule> schedulesByIdMutable;
+    private final WritableKVState<String, List<Schedule>> schedulesByEqualityMutable;
+    private final WritableKVState<Long, List<Schedule>> schedulesByExpirationMutable;
+
+    /**
+     * Create a new {@link WritableScheduleStoreImpl} instance.
+     *
+     * @param states The state to use.
+     */
+    public WritableScheduleStoreImpl(@NonNull final WritableStates states) {
+        super(states);
+        schedulesByIdMutable = states.get(ScheduleServiceImpl.SCHEDULES_BY_ID_KEY);
+        schedulesByEqualityMutable = states.get(ScheduleServiceImpl.SCHEDULES_BY_EQUALITY_KEY);
+        schedulesByExpirationMutable = states.get(ScheduleServiceImpl.SCHEDULES_BY_EXPIRY_SEC_KEY);
```

Comment:
Seems we could compute one or both of these data structures instead of storing them in state. What's the reason for the extra storage cost?